File: Module 20-Advance-sorting/01_mergesort.py
def mergesort(arr):
    if(len(arr)<=1):
        return arr
    arr1 = arr[0:len(arr)//2]
    arr2 = arr[len(arr)//2:len(arr)]
    mergesort(arr1)
    mergesort(arr2)

    mergearray(arr1,arr2,arr)
    return arr


def mergearray(arr1,arr2,arr):
    i,j,k=0,0,0
    while(k<len(arr1)+len(arr2)):
        if(i>=len(arr1)):
            arr[k]=arr2[j]
            k+=1
            j+=1
        elif(j>=len(arr2)):
            arr[k]=arr1[i]
            i+=1
            k+=1
        else:
            if(arr1[i]>=arr2[j]):
                arr[k]=arr2[j]
                k+=1
                j+=1

            else:
                arr[k]=arr1[i]
                i+=1
                k+=1
    return arr


        

        


arr=[2, 1, 23, 9, 5, 8, 7]
mergesort(arr)
print(arr)

File: Module 20-Advance-sorting/01_Mergesort.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> mergeArray(vector<int> arr1, vector<int> arr2) {
    int i = 0, j = 0;
    vector<int> arr;

    while (i < arr1.size() && j < arr2.size()) {
        if (arr1[i] <= arr2[j]) {
            arr.push_back(arr1[i]);
            i++;
        } else {
            arr.push_back(arr2[j]);
            j++;
        }
    }

    // Add leftovers
    while (i < arr1.size()) {
        arr.push_back(arr1[i]);
        i++;
    }
    while (j < arr2.size()) {
        arr.push_back(arr2[j]);
        j++;
    }

    return arr;
}

vector<int> mergeSort(vector<int> arr) {
    if (arr.size() <= 1) return arr;

    int mid = arr.size() / 2;

    vector<int> arr1(arr.begin(), arr.begin() + mid);
    vector<int> arr2(arr.begin() + mid, arr.end());

    arr1 = mergeSort(arr1);
    arr2 = mergeSort(arr2);

    return mergeArray(arr1, arr2);
}

int main() {
    vector<int> arr = {2, 1, 23, 9, 5, 8, 7};
    arr = mergeSort(arr);

    for (int x : arr) cout << x << " ";
    cout << endl;

    return 0;
}
